### PR TITLE
gh-105983: Add private mode support to webbrowser

### PR DESCRIPTION
**Pull request by kr-g**
Support incognito / private browsing.

* Issue: gh-105983

### CODE REVIEW

**Comment by kr-g** [comment]
tested on linux with
- chromium
- firefox
- epiphany

opera opens automatically with a private tab.
konqueror seams not offer such a feature.

File: Lib/webbrowser.py
Comment (sobolevn) [inline Lib/webbrowser.py]: Please, remove unrelated changes: inluding spaces and newlines. It damages the git history for no clear benefit.

Plus, it is hard to review: because I have to get through unrelated changes to find actual ones.

File: Lib/webbrowser.py
Comment (sobolevn) [inline Lib/webbrowser.py]: And things like `'` -> `"` :)

File: Misc/NEWS.d/next/Security/2023-06-22-09-14-45.gh-issue-105983.7osoHj.rst
Comment (sobolevn) [inline Misc/NEWS.d/next/Security/2023-06-22-09-14-45.gh-issue-105983.7osoHj.rst]:
```suggestion
Add support for incognito / private browsing for :func:`webbrowser.open`.
```

**Review by sobolevn — CHANGES_REQUESTED** [review]
Thanks a lot for the feature suggestion. 

I will take another look when you will remove unrelated changes :)

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: i used black for formating the code (PEP08)

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:20]: i used black for formating the code (PEP08).
i will revert the old code base and put my code again.
i would suggest that someone else runs then black (PEP08)

File: Lib/webbrowser.py
Comment (sobolevn) [inline Lib/webbrowser.py:636]: Why do we have `-b` and `-l` added? I am not opposed to this, but again: this feature is not related. 

If you feel like adding this feature is a good idea (it looks like it might be), please open a new issue: explain the use-case and the "why".

But, let's keep this PR simple: only `incognito` mode.

**Review by sobolevn — COMMENTED** [review]
Please, also add tests for browsers that support it. And a doc note about this new feature.

Thank you for the quick response! 👍

**Comment by kr-g** [comment]
i would suggest that someone else will do auto formatting with black (PEP08)

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:636]: it is related to the cmdline interface  when webbroser main() func is called.
they belong also to the newly added incognito method.
therefore it is not a new feature from my perspective.

File: Lib/webbrowser.py
Comment (sobolevn) [inline Lib/webbrowser.py]:
```suggestion
                        "expected 0, 1, 2, or 3, got %s" % new)
```

**Comment by kr-g** [comment]
> Please, also add tests for browsers that support it. And a doc note about this new feature.
> 

i had a look at the tests in Lib/test/test_webbrowser.py and i have really no idea how to test if the browser really opens in private/ incognito mode.

another thing is also because even with new=3 it falls back to new=0 in case remote invocation fails.

see here:
https://github.com/kr-g/cpython/blob/b9079f49b91b45bd439442a9a15fbd02ed6a6a4d/Lib/webbrowser.py#L313

regarding doc note:
see here
https://github.com/kr-g/cpython/blob/9e4ee5958793e90d4544933d9591b689e74a636c/Lib/webbrowser.py#L103

i also saw just right now that here is something wrong with numbering (should be new=2)
https://github.com/kr-g/cpython/blob/9e4ee5958793e90d4544933d9591b689e74a636c/Lib/webbrowser.py#L250

File: Lib/webbrowser.py
Comment (terryjreedy) [reply Lib/webbrowser.py:636]: This is not resolved at all.  These are separate new additions.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Remove stuff unrelated to the issue.

File: Lib/webbrowser.py
Comment (terryjreedy) [inline Lib/webbrowser.py]: This list includes obsolete browsers removed by #102872.

File: Lib/webbrowser.py
Comment (merwok) [inline Lib/webbrowser.py]: an*

File: Lib/webbrowser.py
Comment (merwok) [inline Lib/webbrowser.py]: No need to assign to a variable

File: Lib/webbrowser.py
Comment (merwok) [inline Lib/webbrowser.py]: Bare excepts should be avoided in 99% of cases, they can hide typos and other unintended things.  Catch the most specific error possible.

File: Lib/webbrowser.py
Comment (merwok) [reply Lib/webbrowser.py:636]: Agreed

**Review by merwok — CHANGES_REQUESTED** [review]

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:662]: feel free to split that and follow up

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: i also dislike that type of hard coded stuff.
but i also dont know a better solution here.
anyway that module has a lot of hard code stuff inside.

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: yep

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: i have not checked with ... flake8 (?) ... i have to admit

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: i know

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: https://github.com/kr-g/cpython/commit/333c0e796a783617cf4ffe2ed3cd62a7fa25a31f

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: have you tested how it works?

**Comment by kr-g** [comment]
on my side almost 23pm.
have a good night!
see you tomorrow

File: Lib/webbrowser.py
Comment (merwok) [reply Lib/webbrowser.py]: CPython doesn’t use these tools.  `get(bt)` on the line is enough!

File: Lib/webbrowser.py
Comment (merwok) [reply Lib/webbrowser.py:636]: We do appreciate your contribution here!
As the PR author, you are in the best place to address the review by removing the code added to support `--browser`.
That can be discussed in a separate issue.  But if you don’t have the time, then someone else could complete the PR.

File: Lib/webbrowser.py
Comment (DBJim) [reply Lib/webbrowser.py]: These are the browsers that were removed in 3.12:
"galeon",
"skipstone",
"mosaic",
"grail",

Could you please remove these from the list?

Comment (merwok) [reply Lib/webbrowser.py]: For this PR, the change would be to remove the list.
If another ticket is created to add `--list` option to webbrowser, there the correct list should be added, possibly built dynamically by the `register` function to avoid this kind of issue!

Comment (kr-g) [reply Lib/webbrowser.py]: https://github.com/kr-g/cpython/commit/308e060f0c67510c0af0f39b3418bab756c7e6ce

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: https://github.com/kr-g/cpython/commit/fd0e6a39f196838f5c6a626276e7313eb3d6df6b

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:636]: for me this is a migration topic. the experts do know better what to do here. myself i dont see here in the working loop since i dont have this detailed knowledge of what being best on master code baseline

**Comment by kr-g** [comment]
> i followed the approach of the given tests under Lib/test/test_webbrowser.py and added some more for parameter checking (as far i understood that is how it works).
> 
> but since even the original tests i can not run here on my local machine (3 are failing) i created now a gist rather than releasing it.
> 
> https://gist.github.com/kr-g/95a06df8f0fcf77ad10474724350e921

i created a new issue regarding the failing tests.
https://github.com/python/cpython/issues/106009

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:658]: i just added you all to this repo.
do let me know in case further authorization is missing.

File: Lib/webbrowser.py
Comment (hugovk) [inline Lib/webbrowser.py]: Shall we take this opportunity to f-string it? 
```suggestion
                        f"expected 0, 1, 2, or 3, got {new}")
```

File: Misc/NEWS.d/next/Security/2023-06-22-09-14-45.gh-issue-105983.7osoHj.rst
Comment (hugovk) [inline Misc/NEWS.d/next/Security/2023-06-22-09-14-45.gh-issue-105983.7osoHj.rst:1]: Please move this file from the `Security` directory to `Library`.

**Review by hugovk — COMMENTED** [review]
"Incognito" is Google's name, should we primarily use the generic "private browsing" or "private mode"? We can also mention incognito in docs, as people might search for that term. 

https://en.wikipedia.org/wiki/Private_browsing#Support_in_popular_browsers

File: Misc/NEWS.d/next/Security/2023-06-22-09-14-45.gh-issue-105983.7osoHj.rst
Comment (kr-g) [reply Misc/NEWS.d/next/Security/2023-06-22-09-14-45.gh-issue-105983.7osoHj.rst:1]: https://github.com/kr-g/cpython/commit/e952efe5e8813a68618ecaa273954f7874294637

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py]: i gave also access to you for this repo...
probably pep08 in addition ...

**Comment by kr-g** [comment]
> "Incognito" is Google's name, should we primarily use the generic "private browsing" or "private mode"? We can also mention incognito in docs, as people might search for that term.
> 
> https://en.wikipedia.org/wiki/Private_browsing#Support_in_popular_browsers

in Epiphany it is --incognito-mode

**Comment by kr-g** [comment]
i added now the tests https://github.com/kr-g/cpython/commit/ade902e4f58c0cbc3e126859ade2f5d1be3d7c24
the issue https://github.com/python/cpython/issues/106009 is still open
and i will see what is the outcome on the build/test system.
usually i would not check in a test when parts of it fails on my system
but it might be have no impact since the build/test system is mostlikely configured different

**Comment by kr-g** [comment]
the webbrowser tests passed.
but patchcheck.py failed
https://dev.azure.com/Python/cpython/_build/results?buildId=130810&view=logs&j=256d7e09-002a-52d7-8661-29ee3960640e&t=3d7276d3-4e8d-5309-55ad-fb0b172d9925

but the azure pipeline is not marked as "required"...

**Comment by kr-g** [comment]
```
***patched 1
test_open (__main__.BackgroundBrowserCommandTest.test_open) ... ok
test_register (__main__.BrowserRegistrationTest.test_register) ... ok
...
test_open (__main__.ChromeCommandTest.test_open) ... ok
test_open_incognito (__main__.ChromeCommandTest.test_open_incognito) ... ok
test_open_new (__main__.ChromeCommandTest.test_open_new) ... ok
test_open_new_tab (__main__.ChromeCommandTest.test_open_new_tab) ... ok
...
test_open (__main__.EpiphanyCommandTest.test_open) ... ok
test_open_incognito (__main__.EpiphanyCommandTest.test_open_incognito) ... ok
...
test_open (__main__.MozillaCommandTest.test_open) ... ok
test_open_incognito (__main__.MozillaCommandTest.test_open_incognito) ... ok
test_open_new (__main__.MozillaCommandTest.test_open_new) ... ok
test_open_new_tab (__main__.MozillaCommandTest.test_open_new_tab) ... ok
...

```

fine with me.

**Comment by kr-g** [comment]
for the bot.
I have made the requested changes; please review again

for all others.
can we summarize what needs to be done in addition?
from my side i can not do more.
i gave access to the forked repo if someone needs to work there.
all ideas / hints / splitting / whatever.
please discuss this together, and just do it ... i will watch you doing.
(and learn from that...)

File: Lib/test/test_webbrowser.py
Comment (hugovk) [inline Lib/test/test_webbrowser.py]: Delete this line to fix `make patchcheck` in Azure Pipelines:
```suggestion
```

**Comment by kr-g** [comment]
all improvements send to me here on github (via this collaborating tool/ thing) will be accepted by me without further testing.

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:22]: read at the end. do send proposals. i will accept any.

**Comment by terryjreedy** [comment]
@sobolevn @merwok I leave this to you two as you have worked on the module before.  I have never used it and have other PRs to work on with no other coredevs involved.

File: Lib/webbrowser.py
Comment (ronaldoussoren) [inline Lib/webbrowser.py:109]: Falling back to "open_new" feels like "errors passing silently" to me. Why was this choice made?

**Review by ronaldoussoren — COMMENTED** [review]
In general: this PR does not touch the ``MacOSXOSAScript`` class and cannot because the system interface used by that class doesn't support incognito/private browsing windows (I checked).

The PR does not include a documentation update.

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:129]: not all browsers support that. or some open a private tab by default (like opera) where it would make no sense. for browsers supporting such a feature it is implemented. so there is no "errors passing silently". please provide a use-case where it might fail if you disagree.

**Comment by kr-g** [comment]
> The PR does not include a documentation update.

this is my first PR. 
do please let me know how / where this must be done!

**Comment by kr-g** [comment]
it is possible to fetch the PR to a local repo.

in case you dont know how:

please follow up here https://github.com/progit/progit2/releases -> search for "Pull Request Refs" (in release 2.1.404 it is from page 196 onwards). 

the next section "Pull Requests on Pull Requests" describes how depending changes/ extensions can be made.

so ... for any other browser not covered so far (like e.g. MacOSXOSAScript mentioned by @ronaldoussoren) ... it would be great if the "finder" of that gap could provide also some code here.

Thanks

File: Lib/webbrowser.py
Comment (ronaldoussoren) [reply Lib/webbrowser.py:109]: As a user of the API I'd be disappointed if an URL is opened in regular mode when I explicitly asked to open it in incognito mode.  I'd therefore prefer to raise an exception when incognito mode isn't available.

For use cases where incognito mode is preferred but optional the user can code this themselves, whereas you cannot build my preferred semantics on top of the current implementation.

**Comment by ronaldoussoren** [comment]
> @ronaldoussoren
> 
> > The PR does not include a documentation update.
> 
> this is my first PR. do please let me know how / where this must be done!

The documentation for the webbrowser module is in ``Doc/library/webbrowser.rst `` in the source tree.

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:129]: well. when i look at the function doc and specification. the only thing i can say: "works as designed". 

but having `os.makedirs(..., exist_ok=True)` in mind; i would say: why not?!

you want it; you do it!

but keep it as default to raise no Exception (as it is just right now). since this also follows the new_tab, new_window approach to start a browser with a given url in any case. because this is (probably) mostlikely what comes to the closest what the user expects.

Comment (ronaldoussoren) [reply Lib/webbrowser.py:109]: The two API use cases are different: open_new_tab vs open_new is "just" a different look and feel. "open_incognito" has semantic differences in the browser state (such as not storing the page in long term history and separate cookie  stores).  Opening a page in a regular window when an incognito window was requested can result in unwanted side effects.

I wouldn't mind a keyword argument to enable a fallback to open_new, but it shouldn't be the default. Just like ``exists_ok`` defaults to false in ``os.makedirs``.  Something like ``def open_incognito(url, *, regular_ok=False)``.

Comment (hugovk) [reply Lib/webbrowser.py:109]: I agree with @ronaldoussoren: if a user is asking to keep their private URL secret, we must not "leak" that to the open internet, which could have extremely serious consequences for some people. Let's raise an exception instead.

Regarding Opera: if the user asks for a regular tab and it opens in private tab, that's fine, nothing is leaked.

Comment (kr-g) [reply Lib/webbrowser.py:129]: fine. 2 of 3 (@arhadthedev has not replied so far) voted. 

@hugovk could you please write the missing documentation so that we know how it works and how the function is defined. 

that also would fit into a "documentation first" development flow. 
as soon documentation is available @ronaldoussoren could start here to send his proposal.

i just saw that the issue to this PR was tagged with 3.13 - so - we spent already a lot of time talking. hopefully we not behind schedule now.

Comment (hugovk) [reply Lib/webbrowser.py:109]: We've plenty of time: 3.13 will be released in October 2024 with a May 2024 deadline to get new features in. 

For documentation, something like:

> `webbrowser.open_private(url) `
> Open *url* in a private/incognito window of the default browser, if possible, otherwise raise `RuntimeError`.

@ronaldoussoren has already made the proposal, @kr-g please could you update this PR?

Comment (arhadthedev) [reply Lib/webbrowser.py:109]: I have no preference on whether it should be an exception of a separate method or a permissive argument.

*I can agree with others about the issue in general only because a user who is supposed to get the incognito mode most likely will be unaware of it. So we have a choice between "always show user a warning in advance" vs "let the application gracefully treat a failure by showing an absolutely relevant warning with instructions (or an error if the very fact of page visitation must not be leaked)". Personally, I prefer the second option.*

The exact interface choice is out of my competence.

**Comment by kr-g** [comment]
i updated the documentation regarding my development efforts.
you can add / make changes to this.

**Comment by kr-g** [comment]
i leave it up to you to merge the PR now or extend it and merge later.

or - others requested it already (somehow) - to split this PR (or create a new PR) with the new ideas to address.

**Comment by kr-g** [comment]
@AmirNakhaeii i will check in a few days here again regarding your development. i think your efforts for working into the topic and providing code should be rewarded as well (in the sense of getting part of this PR)

**Comment by AmirNakhaeii** [comment]
> it is possible to fetch the PR to a local repo.
> 
> in case you dont know how:
> 
> please follow up here https://github.com/progit/progit2/releases -> search for "Pull Request Refs" (in release 2.1.404 it is from page 196 onwards).
> 
> the next section "Pull Requests on Pull Requests" describes how depending changes/ extensions can be made.
> 
> so ... for any other browser not covered so far (like e.g. MacOSXOSAScript mentioned by @ronaldoussoren) ... it would be great if the "finder" of that gap could provide also some code here.
> 
> Thanks

I fetched the this PR and I made some changes, but now how can I make the PR for this PR ?

**Comment by kr-g** [comment]
it is in fact a bit complicated. 
i try to do my best!

File: Lib/webbrowser.py
Comment (kr-g) [reply Lib/webbrowser.py:22]: i think this one needs rework.
i checked against the procedure how `register_standard_browsers()` works and got a resulting list with some strange entries. but it seams that those are just aliases for others. so i will remove this list completly

**Comment by hugovk** [comment]
Thanks for the update! Some things still to do (I unresolved the relevant threads):

* [ ] Don't fallback to `open_new`, raise an exception instead
* [ ] Remove `-b` and `-l` options
* [ ] "Incognito" is more of a Google Chrome name, let's primarily name as "private", but can also mention "incognito" to help when searching (I can't seem to find where I earlier commented this...)

**Comment by kr-g** [comment]
> Thanks for the update! Some things still to do (I unresolved the relevant threads):
> 
>     * [ ]  Don't fallback to `open_new`, raise an exception instead
> 
>     * [ ]  Remove `-b` and `-l` options
> 
>     * [ ]  "Incognito" is more of a Google Chrome name, let's primarily name as "private", but can also mention "incognito" to help when searching (I can't seem to find where I earlier commented this...)

@ronaldoussoren @hugovk @arhadthedev please follow up what/ how you want it. from your last reply's it was not clear how exactly it should work. there were some ideas - at least.

**Comment by hugovk** [comment]
@kr-g There's not much point updating this branch from `main` unless the unresolved items are addressed.

File: Lib/webbrowser.py
Comment (merwok) [reply Lib/webbrowser.py:636]: Sorry, I don’t understand «migration».  There are clear requests to remove some code that was added, to keep this PR focused on one thing.  If you don’t want to pursue it, then someone else will have to volunteer.

**Comment by kr-g** [comment]
> @kr-g As mentioned, there's no point updating this PR from `main` unless the other issues are going to be resolved. Otherwise we let's close this, and we can re-open it when those issues have been addressed.

but anyway good to know the intention to kick me out of my own code.

**Comment by AlexWaygood** [comment]
Hi @kr-g, we have no intention of kicking you out of your own code :) However, by keeping a PR open against the CPython repo, you are indicating that this patch is something you are actively seeking to be merged into the `main` branch, and would like CPython maintainers to engage with. If you are not interested in addressing the comments of the reviewers here, that is your right, but you cannot expect for this PR to be kept open indefinitely.

If the PR is closed, the branch will not be deleted, and you will be able to continue using it locally. You will not be kicked out of your own code in any way. However, the PR will no longer show up as an active, open PR against the CPython repo that maintainers should engage with and attempt to review.

It sounds as though you are currently uninterested in addressing the comments of the reviewers here, so I'm going to close this. If I'm incorrect, and you are still interested in having this patch merged into the CPython `main` branch, please let us know, and we can reopen the PR :)

**Comment by kr-g** [comment]
i already said it some months ago that i see the "experts" here in charge to take over those topics because from my point of view thats an integration (or migration) issue to split the PR or rename a feature. i m not an expert and dont know what fits best on cpython baseline. but somehow this turns out to be ... well anyone can read and have an own idea about it.
there is no need to discuss further. also these lines are already too much.

**Comment by AlexWaygood** [comment]
CPython is an open-source project maintained by volunteers, who have limited time to help PRs make it over the finish line. I'm afraid it's the responsibility of the author of the PR to ensure that it's of a quality suitable to be merged, not the responsibility of the CPython core developers. We've given you advice on the changes that would be required to get this into a state where it could be merged; in this instance, that's as much as we can do.

**Pull request closed without merging**
